pkg/providers/amifamily/ami.go

```diff
@@ -126,163 +82,131 @@ func (p *DefaultProvider) List(ctx context.Context, nodeClass *v1beta1.EC2NodeCl
-	return lo.Values(images), nil
-}
-
-type FiltersAndOwners struct {
-	Filters []*ec2.Filter
-	Owners  []string
-}
 
-func GetFilterAndOwnerSets(terms []v1beta1.AMISelectorTerm) (res []FiltersAndOwners) {
 	idFilter := &ec2.Filter{Name: aws.String("image-id")}
-	for _, term := range terms {
+	queries := []AMIQuery{}
+	for _, term := range nodeClass.Spec.AMISelectorTerms {
 		switch {
 		case term.ID != "":
 			idFilter.Values = append(idFilter.Values, aws.String(term.ID))
 		default:
-			elem := FiltersAndOwners{
+			query := AMIQuery{
 				Owners: lo.Ternary(term.Owner != "", []string{term.Owner}, []string{}),
 			}
 			if term.Name != "" {
 				// Default owners to self,amazon to ensure Karpenter only discovers cross-account AMIs if the user specifically allows it.
 				// Removing this default would cause Karpenter to discover publicly shared AMIs passing the name filter.
-				elem = FiltersAndOwners{
+				query = AMIQuery{
 					Owners: lo.Ternary(term.Owner != "", []string{term.Owner}, []string{"self", "amazon"}),
 				}
-				elem.Filters = append(elem.Filters, &ec2.Filter{
+				query.Filters = append(query.Filters, &ec2.Filter{
 					Name:   aws.String("name"),
 					Values: aws.StringSlice([]string{term.Name}),
 				})
 
 			}
 			for k, v := range term.Tags {
 				if v == "*" {
-					elem.Filters = append(elem.Filters, &ec2.Filter{
+					query.Filters = append(query.Filters, &ec2.Filter{
 						Name:   aws.String("tag-key"),
 						Values: []*string{aws.String(k)},
 					})
 				} else {
-					elem.Filters = append(elem.Filters, &ec2.Filter{
+					query.Filters = append(query.Filters, &ec2.Filter{
 						Name:   aws.String(fmt.Sprintf("tag:%s", k)),
 						Values: []*string{aws.String(v)},
 					})
 				}
 			}
-			res = append(res, elem)
+			queries = append(queries, query)
 		}
 	}
 	if len(idFilter.Values) > 0 {
-		res = append(res, FiltersAndOwners{Filters: []*ec2.Filter{idFilter}})
+		queries = append(queries, AMIQuery{Filters: []*ec2.Filter{idFilter}})
+	}
+	return queries, nil
+}
+
+func (p *DefaultProvider) getAMIs(ctx context.Context, queries []AMIQuery) (AMIs, error) {
+	hash, err := hashstructure.Hash(queries, hashstructure.FormatV2, &hashstructure.HashOptions{SlicesAsSets: true})
+	if err != nil {
+		return nil, err
+	}
+	if images, ok := p.cache.Get(fmt.Sprintf("%d", hash)); ok {
+		// Ensure what's returned from this function is a deep-copy of AMIs so alterations
+		// to the data don't affect the original
+		return append(AMIs{}, images.(AMIs)...), nil
 	}
-	return res
+	images := map[uint64]AMI{}
+	for _, query := range queries {
+		if err = p.ec2api.DescribeImagesPagesWithContext(ctx, query.DescribeImagesInput(), func(page *ec2.DescribeImagesOutput, _ bool) bool {
+			for _, image := range page.Images {
+				arch, ok := v1beta1.AWSToKubeArchitectures[lo.FromPtr(image.Architecture)]
+				if !ok {
+					continue
+				}
+				requirementSets := func() []scheduling.Requirements {
```

Comment:
This adds the architecture requirement for an image and then returns those requirements, right? If there are no requirements, just returns the architecture? I'm having a hard time understanding why it needs to be an in-line funciton and why you need the lo.Map() function on L162.

Reply:
If there are no known requirements, this returns a single requirement with the architecture. If there are known requirements, it inserts the architecture requirement to each (via the map) and returns the set. There are definitely different ways to implement this, but this was the cleanest I came up with. This could totally be implemented differently, but using [IIFE](https://en.wikipedia.org/wiki/Immediately_invoked_function_expression) is a personal preference of mine and I have seen it elsewhere in the project. I do think this could fit well as a receiver function on `AMIQuery` though so I'll probably make that change.